virtool/users/sessions.py

```diff
@@ -2,219 +2,220 @@
+        """
+        session_id = await self.create_session_id()
+
+        if user_id and remember:
+            expires_after = timedelta(days=30).total_seconds()
+        elif user_id:
+            expires_after = timedelta(minutes=60).total_seconds()
+        else:
+            expires_after = timedelta(minutes=10).total_seconds()
+
+        new_session = Session(created_at=isoformat(virtool.utils.timestamp()), ip=ip)
+
+        token = None
+
+        if user_id:
+            token, hashed = virtool.utils.generate_key()
+            new_session.authentication = SessionAuthentication(
+                token=hashed, user_id=user_id
+            )
+
+        await self.redis.set(session_id, dumps(new_session), expire=int(expires_after))
+
+        return session_id, new_session, token
+
+    async def create_session_id(self) -> str:
+        """
+        Create a new unique session id.
+
+        :return: a session id
+
+        """
+        session_id = "session_" + secrets.token_hex(32)
+
+        while await self.redis.get(session_id):
+            session_id = "session_" + secrets.token_hex(32)
+
+        return session_id
+
+    async def get(
+        self, session_id: str, session_token: Optional[str] = None
+    ) -> Tuple[Optional[Session], Optional[str]]:
+        """
+        Get a session and token by its id and token.
+
+        If the passed `session_token` is `None`, an unauthenticated session document
+        matching the `session_id` will be returned. If the matching session is
+        authenticated and token is passed, `None` will be returned.
+
+        Will return `(None, None)` if the session doesn't exist or the session id
+        and token do not go together.
+
+        :param session_id: the session id
+        :param session_token: the secure token for an authenticated session
+        :return: the session object and token
+        """
+
+        try:
+            dict_session = json.loads(await self.redis.get(session_id))
+            session = Session(
+                **{
+                    **dict_session,
+                    "created_at": fromisoformat(dict_session["created_at"]),
+                }
+            )
+        except TypeError:
+            return None, None
+
+        if session.authentication is None:
+            return session, None
+
+        if session_token is None:
+            return None, None
+
+        stored_session_token = session.authentication.token
+        hashed_token = hashlib.sha256(session_token.encode()).hexdigest()
+        if stored_session_token == hashed_token:
+            return session, session_token
 
-    :param redis: the application redis client
-    :return: a session id
+        return None, None
 
-    """
-    session_id = "session_" + secrets.token_hex(32)
+    async def replace(
+        self,
+        session_id: str,
+        ip: str,
+        user_id: Optional[str] = None,
+        remember: Optional[bool] = False,
+    ) -> Tuple[str, Session, str]:
```

Comment:
In general I think the class is a good idea, but I changed the control flow to raise an exception if it can't find the session, so returning the session id so no longer needed. (token was already included as part of the session object for authenticated sessions, so I probably should have removed it earlier...)